### PR TITLE
Add manager project

### PR DESCRIPTION
**Pull request by leehuwuj**

### CODE REVIEW

File: src/manager/frontend/src/client/util.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export function getBaseURL(): string {
+  return typeof window !== "undefined" ? window.location.href : "";
```
Comment (leehuwuj) [inline src/manager/frontend/src/client/util.ts:2]: This is working but a bit crunchy, we have it in RAGapp too (using `window.location.origin`) which is broken with reverse proxy

Comment (marcusschiesser) [reply src/manager/frontend/src/client/util.ts:2]: using a deployment variable and if not set fallback to window.location?

File: docker-compose.manager.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+services:
+  admin-dashboard:
+    image: ragapp/admin-dashboard
```
Comment (marcusschiesser) [inline docker-compose.manager.yml:3]: how about just naming this manager?
```suggestion
    image: ragapp/manager
```

File: docker-compose.manager.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    build: src/manager
+    command:
+      - fastapi
+      - run
+      - main.py
+      - --port
+      - '80'
+      - --root-path
+      - /dashboard
```
Comment (marcusschiesser) [inline docker-compose.manager.yml:12]: move to dockerfile?

File: src/manager/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Prepare the requirements stage
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply src/manager/Dockerfile:1]: wrong link above but good idea to use a fastapi standard

File: docker-compose.manager.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    ports:
+      - '8100:80'
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock:ro
```
Comment (leehuwuj) [inline docker-compose.manager.yml:16]: I'm mounting docker.sock directly to the dashboard app with readonly permission, it's working now. But once we need start/stop/remove the container then we should use a dedicated `socat` container for that (https://github.com/alpine-docker/multi-arch-docker-images/tree/master/socat)

File: src/manager/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply src/manager/Dockerfile:1]: where's the FE build?

File: src/manager/app/routers/services.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    updated_at: str | None
+    status: str
+    image: str
+    restart_count: int
+    exposed_ports: dict
+
+    @computed_field  # type: ignore
+    @property
+    def port(self) -> str | None:
+        main_port = self.exposed_ports.get("8000/tcp", [])
```
Comment (marcusschiesser) [inline src/manager/app/routers/services.py:26]: what constant is 8000?

File: src/manager/frontend/public/next.svg
Comment (marcusschiesser) [inline src/manager/frontend/public/next.svg:1]: no need to add next.svg and vercel.svg

File: src/manager/frontend/src/app/layout.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import type { Metadata } from "next";
+import { Inter } from "next/font/google";
+import "./globals.css";
+
+const inter = Inter({ subsets: ["latin"] });
+
+export const metadata: Metadata = {
+  title: "Create Next App",
+  description: "Generated by create next app",
```
Comment (marcusschiesser) [inline src/manager/frontend/src/app/layout.tsx:9]:
```suggestion
  title: "RAGapp - Manager",
  description: "Manager for RAGapp",
```

File: src/manager/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (leehuwuj) [reply src/manager/Dockerfile:17]: Sorry, here is the correct link: https://fastapi.tiangolo.com/deployment/docker/#docker-image-with-poetry
Thanks for the catch, i just added FE build to the Dockerfile

File: src/manager/app/routers/services.py (hunk 6 above)
Comment (leehuwuj) [reply src/manager/app/routers/services.py]: I just removed the port information, we don't need it now.

File: docker-compose.manager.yml [hunk 8]
```diff
@@ -0,0 +1,46 @@
+services:
+  manager:
+    image: ragapp/manager
+    build:
+      context: src/manager
+      args:
+        # Build time variables to set the basePath for static files
+        - APP_PATH=/manager
```
Comment (marcusschiesser) [inline docker-compose.manager.yml:8]: needed? don't we rewrite them on startup?

File: docker-compose.manager.yml [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    entrypoint: ["/app/entrypoint.sh"]
+    command: ["fastapi", "run", "main.py", "--root-path", "/manager"]
```
Comment (marcusschiesser) [inline docker-compose.manager.yml:10]: why do we need to pass root-path here? shouldn't traefik rewrite the URL?
```suggestion
    command: ["fastapi", "run", "main.py", "--root-path", "/manager"]
```

File: docker-compose.manager.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock:ro
+    environment:
+      - APP_PATH=/manager
```
Comment (marcusschiesser) [inline docker-compose.manager.yml:14]: why not `BASE_URL`?
```suggestion
      - BASE_URL=/manager
```

File: src/manager/frontend/src/client/util.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  // If we are in development, use the local backend server
+  if (process.env.ENVIRONMENT === "dev") {
+    return "http://localhost:8000";
+  }
+  // Otherwise, in production, we can either:
+  // - Use the origin URL
+  // - Or use the NEXT_PUBLIC_API_URL environment variable if it is set (suitable for reverse proxy deployment)
+  if (typeof window !== "undefined") {
+    const w = window as any;
+    if (w.ENV && typeof w.ENV.NEXT_PUBLIC_API_URL === "string") {
```
Comment (marcusschiesser) [inline src/manager/frontend/src/client/util.ts:11]: a little bit confusing to me that you call it `NEXT_PUBLIC_API_URL` as env variables starting with `NEXT_PUBLIC` have a special meaning in NextJS

File: docker-compose.manager.yml (hunk 8 above)
Comment (leehuwuj) [reply docker-compose.manager.yml:8]: Ah, this one is now redundant. Sorry for the confusion

File: docker-compose.manager.yml (hunk 9 above)
Comment (leehuwuj) [reply docker-compose.manager.yml:10]: Yes, it should be rewritten by the traefik. Let me remove using rooth-path here

**Comment by leehuwuj** [comment]
PR for RAGapp changes: https://github.com/ragapp/ragapp/pull/118

File: src/manager/frontend/src/client/util.ts [hunk 12]
```diff
@@ -5,11 +5,11 @@ export function getBaseURL(): string {
   }
   // Otherwise, in production, we can either:
   // - Use the origin URL
```
Comment (marcusschiesser) [inline src/manager/frontend/src/client/util.ts:7]: we're not doing this here

File: src/manager/frontend/src/components/sections/sidebar.tsx [hunk 13]
```diff
@@ -7,20 +7,20 @@ export function SideBar() {
         <div className="hidden bg-muted/40 md:block pt-4">
             <div className="flex-1">
                 <nav className="grid items-start px-2 text-sm font-medium lg:px-4">
-                    <Link
+                    <a
```
Comment (marcusschiesser) [inline src/manager/frontend/src/components/sections/sidebar.tsx:10]: any idea why we can't use Link?

**Review by marcusschiesser — COMMENTED** [review]
@leehuwuj just minor things - LGTM - we can use this as base branch for the other changes (so don't merge it yet)

File: src/manager/frontend/src/components/sections/sidebar.tsx (hunk 13 above)
Comment (leehuwuj) [reply src/manager/frontend/src/components/sections/sidebar.tsx:10]: I just feel that the Link is nextjs is a bit complicated, i ran into Link preload issue with exported mode so i  just switched to use <a> instead. (https://github.com/vercel/next.js/discussions/49607)

File: src/manager/frontend/src/client/util.ts (hunk 12 above)
Comment (marcusschiesser) [reply src/manager/frontend/src/client/util.ts:7]: @leehuwuj how about this?

File: deployments/multiple-ragapps/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,125 @@
+services:
```
Comment (marcusschiesser) [inline deployments/multiple-ragapps/docker-compose.yml:1]: add a small README in each deployments folder, to show what is special

File: patch/frontend/app/components/ui/chat/hooks/use-config.ts [hunk 15]
```diff
@@ -0,0 +1,40 @@
+// Patching to use value from window.ENV.BASE_URL
```
Comment (marcusschiesser) [inline patch/frontend/app/components/ui/chat/hooks/use-config.ts:1]: @leehuwuj you can add this file change to the PR that fixes the base URLs in CL

File: Makefile [hunk 16]
```diff
@@ -1,5 +1,5 @@
 export PYTHONPATH := ${PYTHONPATH}:./create_llama/backend
-export CREATE_LLAMA_VERSION=0.1.19
+export CREATE_LLAMA_VERSION=0.1.23
```
Comment (marcusschiesser) [inline Makefile:2]: TODO: add version that fixes missing base URL bugs

File: deployments/multiple-ragapps/docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  traefik:
+    image: "traefik:v3.1"
+    ports:
+      - 80:80
+    labels:
+      - traefik.enable=true
+      - traefik.docker.network=ragapp-network
+      # Default username: admin, password: admin
+      - traefik.http.middlewares.admin-auth.basicauth.users=${USERNAME:-admin}:${HASHED_PASSWORD:-$$apr1$$3gG256rA$$BomHrC/uMhL807k/t7Wd6/}
```
Comment (marcusschiesser) [inline deployments/multiple-ragapps/docker-compose.yml:10]: add to README.md

File: deployments/scripts/create_qdrant_collection.sh [hunk 18]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
```
Comment (marcusschiesser) [inline deployments/scripts/create_qdrant_collection.sh:1]: move both scripts to `single` as they are only used there for now

File: deployments/multiple-ragapps/docker-compose.yml [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+      - --entrypoints.http.address=:80
+      - --accesslog
+      - --log
+      - --api
+    networks:
+      - ragapp-network
+
+  ragapp-cs50:
+    build: ../..
+    image: ragapp/ragapp:latest
```
Comment (marcusschiesser) [inline deployments/multiple-ragapps/docker-compose.yml:25]: didn't use build in my test, but was pulling `ragapp/ragapp:latest`

Comment (marcusschiesser) [reply deployments/multiple-ragapps/docker-compose.yml:25]: Calling `docker-compose up --build` should do the trick - we should probably add it to the readme

File: src/manager/frontend/package.json [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+    "clsx": "^2.1.1",
+    "dotenv": "^16.4.5",
+    "lucide-react": "^0.414.0",
+    "next": "14.2.5",
+    "react": "^18",
+    "react-dom": "^18",
+    "react-query": "^3.39.3",
+    "tailwind-merge": "^2.4.0",
+    "tailwindcss-animate": "^1.0.7",
+    "react-icons": "^5.2.1"
```
Comment (marcusschiesser) [inline src/manager/frontend/package.json:26]: TBD: might just use lucide for icons, e.g. https://lucide.dev/icons/bot for agent

Comment (marcusschiesser) [reply src/manager/frontend/package.json:26]: or alternatively, as `react-icons` includes lucide's icons (see https://react-icons.github.io/react-icons/icons/fi/), just use lucide

Comment (marcusschiesser) [reply src/manager/frontend/package.json:26]: i mean react-icons

**Review by marcusschiesser — APPROVED** [review]

**Pull request merged**